pkg/cloud/services/ec2/launchtemplate_test.go

```diff
@@ -309,3 +436,850 @@ func TestService_LaunchTemplateNeedsUpdate(t *testing.T) {
+			mockScope := mocks.NewMockEC2Scope(mockCtrl)
+			mockEC2Client := mock_ec2iface.NewMockEC2API(mockCtrl)
+			mockSSMClient := mock_ssmiface.NewMockSSMAPI(mockCtrl)
+
+			s := &Service{
+				scope:     mockScope,
+				EC2Client: mockEC2Client,
+				SSMClient: mockSSMClient,
+			}
+
+			tc.expect(mockScope.EXPECT(), mockEC2Client.EXPECT())
+			tc.check(s.GetLaunchTemplateID(tc.launchTemplateName))
+		})
+	}
+}
+
+func TestDeleteLaunchTemplate(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	testCases := []struct {
+		name      string
+		versionID string
+		expect    func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder)
+		check     func(e error)
+	}{
+		{
+			name:      "successfully deleted launch template Id",
+			versionID: "1",
+			expect: func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder) {
+				scope.V(2).Return(klogr.New()).Times(2)
+				m.DeleteLaunchTemplate(gomock.Eq(&ec2.DeleteLaunchTemplateInput{
+					LaunchTemplateId: aws.String("1"),
+				})).Return(&ec2.DeleteLaunchTemplateOutput{}, nil)
+			},
+			check: func(err error) {
+				if err != nil {
+					t.Fatalf("did not expected an error but got %+v", err)
+				}
+			},
+		},
+		{
+			name:      "failed during launch template id deletion",
+			versionID: "1",
+			expect: func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder) {
+				scope.V(2).Return(klogr.New()).Times(1)
+				m.DeleteLaunchTemplate(gomock.Eq(&ec2.DeleteLaunchTemplateInput{
+					LaunchTemplateId: aws.String("1"),
+				})).Return(nil, awserrors.NewFailedDependency("dependency failure"))
+			},
+			check: func(err error) {
+				if err == nil {
+					t.Fatalf("expected an err: %+v, but got nothing", err)
+				}
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if _, err := setupScheme(); err != nil {
+				t.Fatalf("failed during scheme registration with err: %+v", err)
+			}
+
+			mockScope := mocks.NewMockEC2Scope(mockCtrl)
+			mockEC2Client := mock_ec2iface.NewMockEC2API(mockCtrl)
+			mockSSMClient := mock_ssmiface.NewMockSSMAPI(mockCtrl)
+
+			s := &Service{
+				scope:     mockScope,
+				EC2Client: mockEC2Client,
+				SSMClient: mockSSMClient,
+			}
+
+			tc.expect(mockScope.EXPECT(), mockEC2Client.EXPECT())
+			tc.check(s.DeleteLaunchTemplate(tc.versionID))
+		})
+	}
+}
+
+func TestCreateLaunchTemplate(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	var formatTagsInput = func(arg *ec2.CreateLaunchTemplateInput) {
+		sortTags(arg.TagSpecifications[0].Tags)
+
+		for index := range arg.LaunchTemplateData.TagSpecifications {
+			sortTags(arg.LaunchTemplateData.TagSpecifications[index].Tags)
+		}
+	}
+
+	var userData = []byte{1, 0, 0}
+	testCases := []struct {
+		name   string
+		expect func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder)
+		check  func(s string, e error)
+	}{
+		{
```

Comment:
Is it possible to add following scenario here:
Should fail to form launch template data

Reply:
Added a test case `Test_LaunchTemplateDataCreation `